astropy/cosmology/io/tests/test_yaml.py

```diff
@@ -0,0 +1,155 @@
+
+    sig = inspect.signature(yaml_representer)
+    assert len(sig.parameters) == 1
+    assert "tag" in sig.parameters
+    assert sig.parameters["tag"].default is inspect._empty
+
+    # test function `representer`
+    representer = yaml_representer("!astropy.cosmology.flrw.LambdaCDM")
+    assert callable(representer)
+
+    sig = inspect.signature(representer)
+    assert len(sig.parameters) == 2
+    assert "dumper" in sig.parameters
+    assert "obj" in sig.parameters
+    assert sig.parameters["dumper"].default is inspect._empty
+    assert sig.parameters["obj"].default is inspect._empty
+
+    # test the normal method of dumping to YAML
+    yml = dump(Planck18)
+    assert isinstance(yml, str)
+    assert yml.startswith("!astropy.cosmology.flrw.FlatLambdaCDM")
+
+
+def test_yaml_constructor():
+    """Test :func:`~astropy.cosmology.io.yaml.yaml_constructor`."""
+    # test `yaml_representer`
+    assert callable(yaml_constructor)
+
+    sig = inspect.signature(yaml_constructor)
+    assert len(sig.parameters) == 1
+    assert "cls" in sig.parameters
+    assert sig.parameters["cls"].default is inspect._empty
+
+    # test function `constructor`
+    constructor = yaml_constructor(FlatLambdaCDM)
+    assert callable(constructor)
+
+    sig = inspect.signature(constructor)
+    assert len(sig.parameters) == 2
+    assert "loader" in sig.parameters
+    assert "node" in sig.parameters
+    assert sig.parameters["loader"].default is inspect._empty
+    assert sig.parameters["node"].default is inspect._empty
+
+    # it's too hard to manually construct a node, so we only test dump/load
+    # this is also a good round-trip test
+    yml = dump(Planck18)
+    cosmo = load(yml)
+    assert isinstance(cosmo, FlatLambdaCDM)
+    assert cosmo == Planck18
+    assert cosmo.meta == Planck18.meta
+
+
+##############################################################################
+# Test Unified I/O
+
+
+class ToFromYAMLTestMixin(IOTestMixinBase):
+    """
+    Tests for a Cosmology[To/From]Format with ``format="yaml"``.
+    This class will not be directly called by :mod:`pytest` since its name does
+    not begin with ``Test``. To activate the contained tests this class must
+    be inherited in a subclass. Subclasses must define a :func:`pytest.fixture`
+    ``cosmo`` that returns/yields an instance of a |Cosmology|.
+    See ``TestCosmologyToFromFormat`` or ``TestCosmology`` for examples.
+    """
+
+    def test_to_from_yaml_instance(self, cosmo, to_format, from_format):
+        """Test cosmology -> YAML -> cosmology."""
+        # ------------
+        # To YAML
+
+        yml = to_format('yaml')
+        assert isinstance(yml, str)  # test type
+        assert yml.startswith("!astropy.cosmology.")
+
+        # ------------
+        # From YAML
+
+        # tests are different if the last argument is a **kwarg
+        if tuple(cosmo._init_signature.parameters.values())[-1].kind == 4:
+            got = from_format(yml, format="yaml")
```

Comment:
What's going on here? Needs a better comment in any case.

Reply:
See above. This isn't necessary for this test, but in general metadata is handled differently if the cosmology has kwargs. 
https://github.com/astropy/astropy/blob/73c57e16014a11de8ffc7ee75b66ba7b54ed51c2/astropy/cosmology/io/mapping.py#L104-L114
is the reason. Many of the formats route through ``mapping``.

Reply:
But not YAML in the same way because it is currently assumed that no one is messing with the YAML string, adding arguments.

Reply:
I plan on "public" scoping the test suite hopefully for v5.1 release. I've structured the tests so that it should be very easy for 3rd party packages and contributors. Tests for new I/O are a short mixin. Tests for new Cosmologies have a simple inheritance from the parent class tests. 
But I definitely need to increase the density of in-line comments and docstrings and explanations!